core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -152,78 +153,134 @@ public String toString() {
     public MetadataResponse get(String path) {
 
         MetadataResponse metadataResponse = null;
-        String data = null;
         AbortableInputStream abortableInputStream = null;
-        try {
-            String token = getToken();
-            URI uri = URI.create(endpoint + path);
-            HttpExecuteRequest httpExecuteRequest = REQUEST_MARSHALLER.createDataRequest(uri, SdkHttpMethod.GET, token,
-                                                                                tokenTtl);
-            HttpExecuteResponse response = httpClient.prepareRequest(httpExecuteRequest).call();
-            int statusCode = response.httpResponse().statusCode();
-            Optional<AbortableInputStream> responseBody = response.responseBody();
 
-            if (statusCode == HttpURLConnection.HTTP_OK && responseBody.isPresent()) {
-                abortableInputStream = responseBody.get();
-                data = IoUtils.toUtf8String(abortableInputStream);
-                metadataResponse = new MetadataResponse(data);
-            } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
-                throw SdkServiceException.builder()
-                                         .message("The requested metadata at path ( " + path + " ) is not found ").build();
-            } else if (statusCode == HttpURLConnection.HTTP_OK) {
-                throw SdkClientException.builder()
-                                         .message("Response body empty with Status Code " + statusCode).build();
-            } else {
-                throw SdkClientException.builder()
-                                        .message("Instance metadata service returned unexpected status code " + statusCode)
-                                        .build();
+        for (int tries = 1 ; tries <= retryPolicy.numRetries() + 1; tries ++) {
+
+            try {
+                Optional<String> token = getToken();
+                if (token.isPresent()) {
+                    HttpExecuteResponse response = getDataHttpResponse(path, token.get());
+
+                    int statusCode = response.httpResponse().statusCode();
+                    Optional<AbortableInputStream> responseBody = response.responseBody();
+
+                    if (statusCode == 200) {
+                        if (!responseBody.isPresent()) {
+                            throw SdkClientException.builder()
+                                                     .message("Response body empty with Status Code "  + statusCode).build();
+                        }
+                        abortableInputStream = responseBody.get();
+                        String data = IoUtils.toUtf8String(abortableInputStream);
+                        metadataResponse = new MetadataResponse(data);
+                        return metadataResponse;
+                    }
+                    handleException(statusCode, path);
+                }
+                pauseBeforeRetryIfNeeded(tries);
```

Comment:
It may be cleaner if we move `pauseBeforeRetryIfNeeded(tries);` to after finally block so that we can remove `pauseBeforeRetryIfNeeded(tries);` on line 186